esmvalcore/preprocessor/_time.py

```diff
@@ -108,7 +110,107 @@ def extract_time(cube, start_year, start_month, start_day, end_year, end_month,
+            end_date += datetime.timedelta(seconds=1)
+    return end_date
+
+
+def _convert_duration(duration, freq, time_format, reference):
+    duration_unit = cf_units.Unit(f'{freq} since {time_format}',
+                                  calendar=reference.calendar)
+    delta = duration_unit.convert(duration, reference)
+    return delta
+
+
+def _duration_to_date(cube, duration, reference, sign):
+    time_coord = cube.coord('time')
+    reference_format = reference.strftime(time_coord.cell(0).point.format)
+    reference_unit = cf_units.Unit(f'seconds since {reference_format}',
+                                   calendar=time_coord.units.calendar)
+
+    delta = 0
+    if isinstance(duration, isodate.duration.Duration):
+        years = sign * int(duration.years)
+        months = sign * int(duration.months)
+        days = sign * int(duration.days)
+        delta_years = _convert_duration(years, 'years', reference_format,
+                                        reference_unit)
+        delta_months = _convert_duration(months, 'months', reference_format,
+                                         reference_unit)
+        delta_days = _convert_duration(days, 'days', reference_format,
+                                       reference_unit)
+        delta = delta_years + delta_months + delta_days
+    seconds = sign * int(duration.seconds)
+
+    delta += seconds
+    delta = reference_unit.convert(delta, time_coord.units)
+
+    return time_coord.units.num2date(delta)
+
+
+def _extract_datetime(cube, start_datetime, end_datetime):
+    time_coord = cube.coord('time')
+    time_units = time_coord.units
+    if time_units.calendar == '360_day':
+        if start_datetime.day > 30:
+            start_datetime = start_datetime.replace(day=30)
+        if end_datetime.day > 30:
+            end_datetime = end_datetime.replace(day=30)
+
+    t_1 = PartialDateTime(year=int(start_datetime.year),
+                          month=int(start_datetime.month),
+                          day=int(start_datetime.day),
+                          hour=int(start_datetime.hour),
+                          minute=int(start_datetime.minute),
+                          second=int(start_datetime.second))
+
+    t_2 = PartialDateTime(year=int(end_datetime.year),
+                          month=int(end_datetime.month),
+                          day=int(end_datetime.day),
+                          hour=int(end_datetime.hour),
+                          minute=int(end_datetime.minute),
+                          second=int(end_datetime.second))
+
+    constraint = iris.Constraint(time=lambda t: t_1 <= t.point < t_2)
+    cube_slice = cube.extract(constraint)
+    if cube_slice is None:
+        raise ValueError(
+            f"Time slice {start_datetime.strftime('%Y-%m-%d')} "
+            f"to {end_datetime.strftime('%Y-%m-%d')} is outside "
+            f"cube time bounds {time_coord.cell(0)} to {time_coord.cell(-1)}.")
```

Comment:
```suggestion
            f"cube time range {time_coord.cell(0)} to {time_coord.cell(-1)}.")
```
It looks like you're not taking the time bounds into account, are you?